graph/api/src/main/scala/org/apache/spark/graph/api/PropertyGraph.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.graph.api
+
+import org.apache.spark.annotation.Evolving
+import org.apache.spark.sql.{Dataset, Row}
+
+/**
+ * A Property Graph as defined by the openCypher Property Graph Data Model.
+ *
+ * A graph is always tied to and managed by a [[CypherSession]].
+ * The lifetime of a graph is bound by the session lifetime.
+ * The implementation of Cypher in this module corresponds to the Cypher 9 specification
+ * published by the openCypher initiative.
+ *
+ * @see <a href="http://www.opencypher.org/">openCypher project</a>
+ * @see <a href="https://dl.acm.org/citation.cfm?id=3183713.3190657">Property Graph Model</a>
+ * @since 3.0.0
+ */
+@Evolving
+abstract class PropertyGraph {
+
+  /**
+   * The schema (graph type) describes the structure of this graph.
+   *
+   * @since 3.0.0
+   */
+  def schema: PropertyGraphType
+
+  /**
+   * The session in which this graph is managed.
+   *
+   * @since 3.0.0
+   */
+  def cypherSession: CypherSession
+
+  /**
+   * Executes a Cypher query in the session that manages this graph, using this graph as
+   * the input graph.
+   *
+   * @param query Cypher query to execute
+   * @since 3.0.0
+   */
+  def cypher(query: String): CypherResult = cypher(query, Map.empty[String, Any])
+
+  /**
+   * Executes a Cypher query in the session that manages this graph, using this graph as
+   * the input graph.
+   *
+   * @param query      Cypher query to execute
+   * @param parameters parameters used by the Cypher query
+   * @since 3.0.0
+   */
+  def cypher(query: String, parameters: Map[String, Any]): CypherResult =
```

Comment:
See my comments on overloading `cypher`.

Reply:
I'd like to address this in a follow up PR.